### DISCUSSION TITLE
Look up system dependencies (moc, replica, ic-starter) at runtime

### DISCUSSION THREAD

**Discussion by nmattia** [dfinity/sdk#2625] [Feature Requests]
Instead of bundling dependencies in `dfx` itself, those dependencies could be searched for and downloaded only when necessary.  This would solve the frustration expressed by @paulyoung [on the forum](https://forum.dfinity.org/t/external-contributions-to-the-sdk-are-blocked-by-private-dependencies/11784/15) by not needing to download dependencies during the build.

In general there are drawbacks to embedding dependencies. It is not considered good practice to download artifacts during the build, and in this case bloats `dfx` a fair bit. Instead, much like e.g. `cargo` does, dependencies could simply be checked at runtime. In the case of `dfx`, we could even offer to download and install those dependencies (which `cargo` doesn't do, i.e. it does not download `rustc`).

This would also make `dfx` a fair bit lighter, because optional dependencies (like `moc`) may never have to be downloaded (e.g. if the developer uses rust but doesn't use motoko).

**Comment by paulyoung** [comment]
From what I remember, the original motivation for this was to address concerns with the first-time developer experience.

We didn’t want new developers thinking they had everything they needed by having already downloaded the SDK and then later finding that they needed an internet connection when they didn’t have one.

I’m not sure if that’s a strong enough argument but I thought the context might be useful.
